### PR TITLE
PatchParameter

### PR DESCRIPTION
**Pull request by tardyp**
Initial design to allow sending patch via the force scheduler UI.

For now, we add the necessary blocks to the force framework, and make sure a patch is properly applied with an integration test.

UI, relnotes, and docs will be in followup PRs (to keep things reviewable).

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the `master/buildbot/newsfragment` directory (and read the `README.txt` in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/test/util/scheduler.py [hunk 1]
```diff
@@ -204,7 +204,7 @@ def fake_addBuildsetForSourceStampsWithDefaults(self, reason, sourcestamps=None,
         self.assertIsInstance(sourcestamps, list)
 
         def sourceStampKey(sourceStamp):
-            return tuple(sorted(sourceStamp.values()))
```
Comment (tardyp) [inline master/buildbot/test/util/scheduler.py]: values() is not reliable order, so we only order by codebase, which should be unique and defined in the sourcestamps

**Comment by tardyp** [comment]
Sorry, its hard to make it small, and splitting the UI doesn't help that much. 
the UI part is actually pretty small..

File: master/docs/manual/cfg-schedulers.rst [hunk 2]
```diff
@@ -1012,8 +1012,15 @@ All parameter types have a few common arguments:
 
     If this is true, then an error will be shown to user if there is no input in this field
 
+``maxsize`` (optional; default: None)
+
+    The maximum size of a field.
```
Comment (seankelly) [inline master/docs/manual/cfg-schedulers.rst]: What's the unit? Bytes?

File: master/docs/manual/cfg-schedulers.rst [hunk 3]
```diff
@@ -1223,6 +1232,52 @@ This is a parameter group to specify a sourcestamp for a given codebase.
...
+
+
+.. bb:sched:: FileParameter
+
+FileParameter
+#############
+
+This parameter allows the user to upload a file to a build. 
+The user can either write some text to a text area, or select a file from the browser.
+Note that the file is then stored inside a property, so a ``maxsize`` of 10 mega bytes has been set.
```
Comment (seankelly) [inline master/docs/manual/cfg-schedulers.rst]: "Megabytes" is one word.

**Pull request merged**